### ISSUE TITLE
Can types that implement both IFormattable and IParsable<T> be IXunitSerializable by default?

### ISSUE THREAD

**Issue by rabuckley** [xunit/xunit#3064]
Often, the types I want to pass in to theories implement `IFormattable` (generally via `ISpanFormattable`) and `IParsable<TSelf>` (generally via `ISpanParsable<TSelf>`).

The interfaces are the canonical way to express serializability to and from string, and are used in [ASP.NET Core route binding](https://learn.microsoft.com/en-us/aspnet/core/mvc/models/model-binding?view=aspnetcore-9.0#bind-with-iparsablettryparse) and [string interpolation](https://github.com/dotnet/runtime/blob/0d55c5bb4f5342e2a880fe62a0202195fecdbe5c/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/DefaultInterpolatedStringHandler.cs#L247-L286), to name a couple of uses.

To get these to play nicely with xunit serialization, I've written the following `IXunitSerializer`

```cs
/// <summary>
/// An adapter to make types that implement <see cref="IFormattable"/> and <see cref="IParsable{T}"/> serializable by xunit.
/// </summary>
/// <typeparam name="T"></typeparam>
public sealed class FormattableAndParsableSerializer<T> : IXunitSerializer
    where T : IFormattable, IParsable<T>
{
    public object Deserialize(Type type, string serializedValue)
    {
        return T.Parse(serializedValue, CultureInfo.InvariantCulture);
    }

    public bool IsSerializable(Type type, object? value)
    {
        return typeof(T) == type && value is T;
    }

    public string Serialize(object value)
    {
        ArgumentNullException.ThrowIfNull(value);
        return ((T)value).ToString(default, CultureInfo.InvariantCulture);
    }
}
```

and for each type I want to use I add (using [IPAddress](https://learn.microsoft.com/en-us/dotnet/api/system.net.ipaddress?view=net-9.0) as an example which isn't built in but does implement the relevant interfaces),

```cs
[assembly: RegisterXunitSerializer(typeof(FormattableAndParsableSerializer<IPAddress>), typeof(IPAddress))]
```

Could xunit have this ability built in? If so, many of the explicit [built in conversions](https://github.com/xunit/xunit/blob/fa81ceac0f280a65d43ec849f6f12bfff811c7f1/src/xunit.v3.common/Serialization/SerializationHelper.cs#L94-L122) would "just light up" - as would many other types.

I don't know enough about xunit and whether or not this would cause breaks to know if this would be feasible, but I thought I'd just ask. Thanks!

**Comment by bradwilson** [comment]
I was curious about this, so I made the serializer non-generic:

```csharp
public sealed class FormattableAndParsableSerializer : IXunitSerializer
{
    public object Deserialize(Type type, string serializedValue)
    {
        // We need to look for the Parse method, and while it's definitely static, it might be public
        // or private, depending on whether the type wanted to hide the implementation of IParseable
        var parse =
            type.GetMethod("Parse", BindingFlags.Static | BindingFlags.Public, [typeof(string), typeof(IFormatProvider)]) ??
            type.GetMethod($"System.IParsable<{type.FullName}>.Parse", BindingFlags.Static | BindingFlags.NonPublic, [typeof(string), typeof(IFormatProvider)]) ??
            throw new InvalidOperationException($"Could not find Parse method for IParsable<{type.FullName}>");

        return
            parse.Invoke(null, [serializedValue, CultureInfo.InvariantCulture])
            ?? throw new InvalidOperationException($"Call to IParsable<{type.FullName}>.Parse(\"{serializedValue}\") returned null");
    }

    public bool IsSerializable(Type type, object? value) =>
        // We already know it's IFormattable, but we need to make sure it's also IParsable<T>
        typeof(IParsable<>).MakeGenericType(type).IsAssignableFrom(type);

    public string Serialize(object value) =>
        ((IFormattable)value).ToString(default, CultureInfo.InvariantCulture);
}
```

and registered it:

```csharp
[assembly: RegisterXunitSerializer(typeof(FormattableAndParsableSerializer), typeof(IFormattable))]
```

and it works for any type that implements `IFormattable` and `IParsable<T>`:

```csharp
var serialized = SerializationHelper.Instance.Serialize(new IPAddress([127, 0, 0, 1]));
var deserialized = SerializationHelper.Instance.Deserialize(serialized);

var address = Assert.IsType<IPAddress>(deserialized);
Assert.Equal("127.0.0.1", address.ToString());
```

When I tried to serialize something that was only `IFormattable`, that uncovered a bug where it successfully serialized (it should not have, since the IsSerializable call should fail). So I set a breakpoint on that, and it's never called. Makes sense, given the behavior, but clearly this is a bug.

So I'm off to fix that bug first. 😂

**Comment by bradwilson** [comment]
I've updated IXunitSerializer.IsSerializable to return an error message so that calls to SerializationHelper.Serialize will generate appropriate messages to help the end user.

This makes the non-generic serializer look like this:

```csharp
public sealed class FormattableAndParsableSerializer : IXunitSerializer
{
    public object Deserialize(Type type, string serializedValue)
    {
        // We need to look for the Parse method, and while it's definitely static, it might be public
        // or private, depending on whether the type wanted to hide the implementation of IParseable
        var parse =
            type.GetMethod("Parse", BindingFlags.Static | BindingFlags.Public, [typeof(string), typeof(IFormatProvider)]) ??
            type.GetMethod($"System.IParsable<{type.FullName}>.Parse", BindingFlags.Static | BindingFlags.NonPublic, [typeof(string), typeof(IFormatProvider)]) ??
            throw new InvalidOperationException($"Could not find Parse method for IParsable<{type.FullName}>");

        return
            parse.Invoke(null, [serializedValue, CultureInfo.InvariantCulture])
            ?? throw new InvalidOperationException($"Call to IParsable<{type.FullName}>.Parse(\"{serializedValue}\") returned null");
    }

    public bool IsSerializable(Type type, object? value, [NotNullWhen(false)] out string? failureReason)
    {
        var isParsable = false;

        // We wrap this in a try/catch because instantiating the generic may throw because
        // of constraint violations.
        try
        {
            isParsable = typeof(IParsable<>).MakeGenericType(type).IsAssignableFrom(type);
        }
        catch { }

        if (!isParsable)
        {
            failureReason = $"Type '{type.FullName}' must implement both IFormattable and IParsable<> to be serialized";
            return false;
        }

        failureReason = null;
        return true;
    }

    public string Serialize(object value) =>
        ((IFormattable)value).ToString(default, CultureInfo.InvariantCulture);
}
```

And now I see errors when trying to serialize things that aren't both:

![image](https://github.com/user-attachments/assets/4b90d915-2dc8-435a-9708-faa52d223059)

**Comment by bradwilson** [comment]
Available in v3 `0.7.0-pre.18` https://xunit.net/docs/using-ci-builds
